Profitable Apps for Google Play and the App Store

This project focuses on isolating specific apps in order to gain more information on how ads in an app can specifically generate revenue for a company. In this, I will be examining free apps and user behaviour in order to gain a better understanding of how to better generate ad revenue.

First, the data needs to be read in. There is a data set of about 10,000 apps from the Google Play store and one of about 7,000 apps from the App Store. 

In [1]:
opened_file_apple = open('AppleStore.csv')
opened_file_google = open('googleplaystore.csv')
from csv import reader
read_file_apple = reader(opened_file_apple)
read_file_google = reader(opened_file_google)

apps_data_apple = list(read_file_apple)
apps_data_google = list(read_file_google)

android_header = apps_data_google[0]
android = apps_data_google[1:]

apple_header = apps_data_apple[0]
apple = apps_data_apple[1:]
print(len(apple))
print(len(android))

7197
10841


Below an explore_data function is defined. This takes each seperate app in the list of apple and android apps and then puts another line in between each app so we can view the data easier. It then possibly prints the amount of data in the set by displaying the number of rows and columns.

In [2]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:We concluded that taking a popular book (perhaps a more recent book) and turning it into an app could be profitable for both the Google Play and the App Store markets. The markets are already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))
        

Since the title rows were removied, the sets can now be explored with the function. The apple store has 7197 apps with 16 unique fields to describe the app. The android store has 10841 apps with 13 unique fields.

In [3]:
explore_data(apple, 0, 3, rows_and_columns = True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


In [4]:
explore_data(android, 0, 3, rows_and_columns = True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


When row 10473 index entry in the android store app list is inspected, it can be seen that there is an error in the info, one of the fields is empty. This must therefore be deleted.

In [27]:
print(apps_data_google[10473])

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


Now there is one less entry in the list length. Now the relevant data can be dealt with.

In [28]:
print(len(android))
del android[10472]
print(len(android))

10840
10839


Now unique apps will be checked for in the App Store and the Google Play store. This way any duplicates can be deleted.

In [7]:
unique_app_names_apple = []
duplicate_app_names_apple = []
for row in apple:
    app_name = row[0]
    if app_name in unique_app_names_apple:
        duplicate_app_names_apple.append(app_name)
    else:
        unique_app_names_apple.append(app_name)
        
for row in apple:
    name = row[0]
    if name == 'Instagram':
        print(row)
        
print("Number of unique apps: " + str(len(unique_app_names_apple)))
print("Number of duplicate apps: " + str(len(duplicate_app_names_apple)))

Number of unique apps: 7197
Number of duplicate apps: 0


In [8]:
unique_app_names_android = []
duplicate_app_names_android = []
for row in android:
    app_name = row[0]
    if app_name in unique_app_names_android:
        duplicate_app_names_android.append(app_name)
    else:
        unique_app_names_android.append(app_name)
        
for row in android:
    name = row[0]
    if name == 'Instagram':
        print(row)
        
print("Number of unique apps: " + str(len(unique_app_names_android)))
print("Number of duplicate apps: " + str(len(duplicate_app_names_android)))

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
Number of unique apps: 9659
Number of duplicate apps: 1181


Its fairly obvious from this output that there are duplicates in the dataset. Purging is needed. Below is the expected length of the list with only unique apps in the Google Play store.

In [9]:
print("Expected length", len(android) - len(duplicate_app_names_android))

Expected length 9659


To decide between which of the duplicate apps to keep, the ones with the most ratings will be kept.

In [29]:
reviews_max = {}

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    
    elif name not in reviews_max:
        reviews_max[name] = n_reviews
        
print("Check for length of dictionary, should be 9659: ", len(reviews_max))


##We now have the max review dictionary values, we can go in and add the actual apps
##now with the relevant information.

Check for length of dictionary, should be 9659:  9659


In [11]:
android_clean = []
android_already_added = []

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if (reviews_max[name] == n_reviews) and (name not in android_already_added):
        android_clean.append(app)
        android_already_added.append(name)
print("Check for length of clean android list now, should be same:", len(android_clean))

Check for length of clean android list now, should be same: 9659


Now, any non english app will be deleted from this since I'm only concerned with english apps.

In [12]:
def english_detector_1(some_string):
    for character in some_string:
        character_number = ord(character)
        if character_number > 127:
            return False
    return True


#test    
print("Instagram result: ", english_detector_1(some_string = "Instagram"))
print("foreign app result: ", english_detector_1(some_string = '爱奇艺PPS -《欢乐颂2》电视剧热播'))

Instagram result:  True
foreign app result:  False


However, this does not cover all cases as some apps might use emojis in their name or other symbols that do not cover the entire number line or english alphabet. So I will edit the function to return false only if there are more than three characters that fall out of the ASCII range (0-127). Otherwise it will return true.

In [13]:
def english_detector_2(some_string):
    foreign_character_counter = 0
    for character in some_string:
        character_number = ord(character)
        
        if character_number > 127:
            foreign_character_counter += 1
    if foreign_character_counter > 3:
        return False
    else:
        return True
        
#test        
print("Instachat result: ", english_detector_2(some_string = "Instachat 😜"))
print("foreign app result: ", english_detector_2(some_string = '爱奇艺PPS -《欢乐颂2》电视剧热播'))

Instachat result:  True
foreign app result:  False


Now that there is a functional english detector for the app names, it willb be applied to these to the two data sets for google and apple. I will go through the data sets of the apps and make a english list of lists.

In [14]:
english_android = []
english_apple = []

print("length of apple list: ", len(apple))
print("length of android list: ", len(android_clean))
for app in android_clean:
    name = app[0]
    if english_detector_2(name):
        english_android.append(app)
        
for app in apple:
    name = app[1]
    if english_detector_2(name):
        english_apple.append(app)
        
print("length of english apple list: ", len(english_apple))
print("length of english android list: ", len(english_android))

length of apple list:  7197
length of android list:  9659
length of english apple list:  6183
length of english android list:  9614


There are now two relatively cleaner app lists in both the ios store and the google store. No duplicates and english apps.

Since this project only is concerned with free apps, I must get rid of apps that are priced in boths stores.

In [15]:
android_final = []
apple_final = []

for app in english_android:
    price  = app[7]
    if price == "0":
        android_final.append(app)
print(len(android_final))

for app in english_apple:
    price  = app[4]
    if price == "0.0":
        apple_final.append(app)
print(len(apple_final))
    
    

8864
3222


Now that I have narrowed down our apps to non duplicate, free, english apps, we can finally begin our analysis.

The aim is to determine the kinds of apps that are likely to attract more users since the revenue we acquire is based on the number of people using our apps. Since the goal is to have this app be on both the Apple and Google play store, we want to analyze both data sets so it can be successful in both markets. Therefore, app profiles that are successful in both markets need to be found. To start this, lets get a sense of the genres that are most common in each market.

In [16]:
print(android_header)
print(apple_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


In [17]:
def freq_table(dataset, index):
    frequency_table = {}
    for app in dataset:
        index_value = app[index]
        if index_value in frequency_table:
            frequency_table[index_value] += 1
        else: 
            frequency_table[index_value] = 1
    for app in frequency_table:
        frequency_table[app] = frequency_table[app] / len(dataset)
        
    return frequency_table

In [18]:
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [19]:
display_table(android_final, 9)
##genres google

Tools : 0.08449909747292418
Entertainment : 0.06069494584837545
Education : 0.05347472924187725
Business : 0.04591606498194946
Productivity : 0.03892148014440433
Lifestyle : 0.03892148014440433
Finance : 0.03700361010830325
Medical : 0.03531137184115524
Sports : 0.03463447653429603
Personalization : 0.03316787003610108
Communication : 0.032378158844765345
Action : 0.03102436823104693
Health & Fitness : 0.030798736462093863
Photography : 0.02944494584837545
News & Magazines : 0.027978339350180504
Social : 0.026624548736462094
Travel & Local : 0.023240072202166066
Shopping : 0.022450361010830325
Books & Reference : 0.021435018050541516
Simulation : 0.020419675090252706
Dating : 0.01861462093862816
Arcade : 0.018501805054151624
Video Players & Editors : 0.017712093862815883
Casual : 0.01759927797833935
Maps & Navigation : 0.013989169675090252
Food & Drink : 0.012409747292418772
Puzzle : 0.01128158844765343
Racing : 0.009927797833935019
Role Playing : 0.009363718411552346
Libraries & Demo 

In [20]:
display_table(android_final, 1)
##category google

FAMILY : 0.18907942238267147
GAME : 0.09724729241877256
TOOLS : 0.08461191335740072
BUSINESS : 0.04591606498194946
LIFESTYLE : 0.039034296028880866
PRODUCTIVITY : 0.03892148014440433
FINANCE : 0.03700361010830325
MEDICAL : 0.03531137184115524
SPORTS : 0.03395758122743682
PERSONALIZATION : 0.03316787003610108
COMMUNICATION : 0.032378158844765345
HEALTH_AND_FITNESS : 0.030798736462093863
PHOTOGRAPHY : 0.02944494584837545
NEWS_AND_MAGAZINES : 0.027978339350180504
SOCIAL : 0.026624548736462094
TRAVEL_AND_LOCAL : 0.023352888086642598
SHOPPING : 0.022450361010830325
BOOKS_AND_REFERENCE : 0.021435018050541516
DATING : 0.01861462093862816
VIDEO_PLAYERS : 0.017937725631768955
MAPS_AND_NAVIGATION : 0.013989169675090252
FOOD_AND_DRINK : 0.012409747292418772
EDUCATION : 0.011620036101083033
ENTERTAINMENT : 0.009589350180505414
LIBRARIES_AND_DEMO : 0.009363718411552346
AUTO_AND_VEHICLES : 0.009250902527075812
HOUSE_AND_HOME : 0.008235559566787004
WEATHER : 0.008009927797833934
EVENTS : 0.0071074007

In [21]:
display_table(apple_final, -5)
##prime genre apple

Games : 0.5816263190564867
Entertainment : 0.07883302296710118
Photo & Video : 0.04965859714463067
Education : 0.03662321539416512
Social Networking : 0.032898820608317815
Shopping : 0.0260707635009311
Utilities : 0.025139664804469275
Sports : 0.021415270018621976
Music : 0.020484171322160148
Health & Fitness : 0.020173805090006207
Productivity : 0.01738050900062073
Lifestyle : 0.015828677839851025
News : 0.01334574798261949
Travel : 0.012414649286157667
Finance : 0.0111731843575419
Weather : 0.008690254500310366
Food & Drink : 0.008069522036002483
Reference : 0.00558659217877095
Business : 0.005276225946617008
Book : 0.004345127250155183
Navigation : 0.00186219739292365
Medical : 0.00186219739292365
Catalogs : 0.0012414649286157666


The games category in the apple store is by far the most popular and more than half of all the free, english apps. 
The google play store is much more varied. Family, games, and tools are the most popular categories but not by nearly as much. Also, tools and entertainment are the most popular genres in the google play store. This could also just speak to how google organizes its app store.
The best bet would be a free game.

Moving on, I'll investigate which kind of apps are the most popular with users. This can be done with installs from the google store and rating_count_tot from the apple store. I will calculate the average number of user ratings per app genre on the App Store.

By step, I will:
isolate the apps of each genre,
sum up the user ratings for the apps of that genre,
then divide the sum by the number of apps belonging to that genre

This will be done using nested loops.

In [22]:
prime_genre_apple = freq_table(apple_final, -5)

for genre in prime_genre_apple:
    total = 0
    len_genre = 0
    
    for app in apple_final:
        genre_app = app[-5]
        
        if genre_app == genre:
            user_ratings = float(app[5])
            total += user_ratings
            len_genre += 1
    avg_user_number = total / len_genre
    print("Genre: ", genre, 
          "   Average number of user ratings: "
          , avg_user_number)

Genre:  Catalogs    Average number of user ratings:  4004.0
Genre:  Medical    Average number of user ratings:  612.0
Genre:  Health & Fitness    Average number of user ratings:  23298.015384615384
Genre:  Shopping    Average number of user ratings:  26919.690476190477
Genre:  Education    Average number of user ratings:  7003.983050847458
Genre:  Weather    Average number of user ratings:  52279.892857142855
Genre:  Lifestyle    Average number of user ratings:  16485.764705882353
Genre:  Music    Average number of user ratings:  57326.530303030304
Genre:  Business    Average number of user ratings:  7491.117647058823
Genre:  Food & Drink    Average number of user ratings:  33333.92307692308
Genre:  Utilities    Average number of user ratings:  18684.456790123455
Genre:  Social Networking    Average number of user ratings:  71548.34905660378
Genre:  Travel    Average number of user ratings:  28243.8
Genre:  Photo & Video    Average number of user ratings:  28441.54375
Genre:  Navigatio

From this table, the navigation genre has the most reviews.

In [23]:
category_android = freq_table(android_final, 1)

In [25]:
for category in category_android:
    total = 0
    len_category = 0
    for app in android_final:
        category_app = app[1]
        if category_app == category:
            installs = app[5]
            installs = installs.replace('+', '')
            installs = installs.replace(',', '')
            installs = float(installs)
            total += installs
            len_category += 1
    avg_number_installs = total / len_category
    print(category, ":", avg_number_installs)

FAMILY : 3695641.8198090694
COMMUNICATION : 38456119.167247385
PERSONALIZATION : 5201482.6122448975
ENTERTAINMENT : 11640705.88235294
NEWS_AND_MAGAZINES : 9549178.467741935
TOOLS : 10801391.298666667
BEAUTY : 513151.88679245283
AUTO_AND_VEHICLES : 647317.8170731707
WEATHER : 5074486.197183099
ART_AND_DESIGN : 1986335.0877192982
EVENTS : 253542.22222222222
PARENTING : 542603.6206896552
FINANCE : 1387692.475609756
VIDEO_PLAYERS : 24727872.452830188
MAPS_AND_NAVIGATION : 4056941.7741935486
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
EDUCATION : 1833495.145631068
MEDICAL : 120550.61980830671
BOOKS_AND_REFERENCE : 8767811.894736841
PHOTOGRAPHY : 17840110.40229885
DATING : 854028.8303030303
BUSINESS : 1712290.1474201474
HEALTH_AND_FITNESS : 4188821.9853479853
FOOD_AND_DRINK : 1924897.7363636363
TRAVEL_AND_LOCAL : 13984077.710144928
SPORTS : 3638640.1428571427
HOUSE_AND_HOME : 1331540.5616438356
SOCIAL : 23253652.127118643
COMICS : 817657.2727272727
LIBRARIES_AND_DEMO : 638503.73

On average communication apps have the most installs.

In this project, data was analyzed about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.

From this, a free mobile game or a navigation / communication app would be a safe type of app to make with ad revenue in mind.